internal/helper/converter/converter.go

```diff
@@ -0,0 +1,126 @@
+package converter
+
+import (
+	"encoding/json"
+	"reflect"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type APIModel interface {
+	MarshalBinary() ([]byte, error)
+	UnmarshalBinary(b []byte) error
+}
+
+type TFSchemaModelConverter[T APIModel] struct {
+	TFModelMap *BlockToStruct
+}
+
+const (
+	arrIndexExceededMsg = "array index exceeded"
+	arrIsNilMsg         = "array is nil"
+)
+
+// ### Public Funcs ###.
+
+func (converter *TFSchemaModelConverter[T]) ConvertTFSchemaToAPIModel(data *schema.ResourceData, buildOnlyKeys []string) (modelPtr T) {
```

Comment:
Can you add a description to the public methods?

Reply:
yes sure

Reply:
Done